File: src/algorithms/linked-lists/insertion-deletion/remove-duplicates-sorted/sources/remove-duplicates-sorted_test.rs
include!("remove-duplicates-sorted.rs");

fn build_list(values: &[i32]) -> Option<Box<ListNode>> {
    let mut head: Option<Box<ListNode>> = None;
    for &val in values.iter().rev() {
        head = Some(Box::new(ListNode { value: val, next: head }));
    }
    head
}

fn list_to_vec(mut head: Option<Box<ListNode>>) -> Vec<i32> {
    let mut result = Vec::new();
    while let Some(node) = head {
        result.push(node.value);
        head = node.next;
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_removes_consecutive_duplicates() {
        let list = build_list(&[1, 1, 2, 3, 3, 3, 4, 5, 5]);
        assert_eq!(list_to_vec(remove_duplicates_sorted(list)), vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_no_duplicates_unchanged() {
        let list = build_list(&[1, 2, 3, 4, 5]);
        assert_eq!(list_to_vec(remove_duplicates_sorted(list)), vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_all_same_values() {
        let list = build_list(&[7, 7, 7, 7]);
        assert_eq!(list_to_vec(remove_duplicates_sorted(list)), vec![7]);
    }

    #[test]
    fn test_empty_list() {
        assert_eq!(remove_duplicates_sorted(None), None);
    }

    #[test]
    fn test_single_element() {
        let list = build_list(&[5]);
        assert_eq!(list_to_vec(remove_duplicates_sorted(list)), vec![5]);
    }

    #[test]
    fn test_two_element_duplicates() {
        let list = build_list(&[3, 3]);
        assert_eq!(list_to_vec(remove_duplicates_sorted(list)), vec![3]);
    }

    #[test]
    fn test_two_different_elements() {
        let list = build_list(&[1, 2]);
        assert_eq!(list_to_vec(remove_duplicates_sorted(list)), vec![1, 2]);
    }

    #[test]
    fn test_mixed_run_lengths() {
        let list = build_list(&[1, 2, 2, 3, 3, 3, 4]);
        assert_eq!(list_to_vec(remove_duplicates_sorted(list)), vec![1, 2, 3, 4]);
    }
}

File: src/algorithms/linked-lists/insertion-deletion/remove-duplicates-sorted/__tests__/remove-duplicates-sorted_test.py
import importlib
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "sources"))

module = importlib.import_module("remove-duplicates-sorted")
remove_duplicates_sorted = module.remove_duplicates_sorted
ListNode = module.ListNode


def build_list(values):
    head = None
    for val in reversed(values):
        head = ListNode(val, head)
    return head


def list_to_array(head):
    result = []
    current = head
    while current is not None:
        result.append(current.value)
        current = current.next
    return result


def test_removes_consecutive_duplicates():
    result = remove_duplicates_sorted(build_list([1, 1, 2, 3, 3, 3, 4, 5, 5]))
    assert list_to_array(result) == [1, 2, 3, 4, 5]


def test_no_duplicates_unchanged():
    result = remove_duplicates_sorted(build_list([1, 2, 3, 4, 5]))
    assert list_to_array(result) == [1, 2, 3, 4, 5]


def test_all_same_values():
    result = remove_duplicates_sorted(build_list([7, 7, 7, 7]))
    assert list_to_array(result) == [7]


def test_empty_list():
    result = remove_duplicates_sorted(None)
    assert result is None


def test_single_element():
    result = remove_duplicates_sorted(build_list([5]))
    assert list_to_array(result) == [5]


def test_two_element_duplicates():
    result = remove_duplicates_sorted(build_list([3, 3]))
    assert list_to_array(result) == [3]


def test_two_different_elements():
    result = remove_duplicates_sorted(build_list([1, 2]))
    assert list_to_array(result) == [1, 2]


def test_mixed_run_lengths():
    result = remove_duplicates_sorted(build_list([1, 2, 2, 3, 3, 3, 4]))
    assert list_to_array(result) == [1, 2, 3, 4]


if __name__ == "__main__":
    test_removes_consecutive_duplicates()
    test_no_duplicates_unchanged()
    test_all_same_values()
    test_empty_list()
    test_single_element()
    test_two_element_duplicates()
    test_two_different_elements()
    test_mixed_run_lengths()
    print("All tests passed.")
